nengo/builder/tests/test_optimizer.py

```diff
@@ -0,0 +1,142 @@
+import numpy as np
+from numpy.testing import assert_almost_equal
+import pytest
+
+import nengo
+from nengo.builder.optimizer import SigMerger
+from nengo.builder.signal import Signal
+from nengo.spa.tests.test_thalamus import thalamus_net
+from nengo.tests.test_learning_rules import learning_net
+
+
+def test_sigmerger_check():
+    # 0-d signals
+    assert SigMerger.check([Signal(0), Signal(0)])
+    assert not SigMerger.check([Signal(0), Signal(1)])
+
+    # compatible along first axis
+    assert SigMerger.check(
+        [Signal(np.empty((1, 2))), Signal(np.empty((2, 2)))])
+
+    # compatible along second axis
+    assert SigMerger.check(
+        [Signal(np.empty((2, 1))), Signal(np.empty((2, 2)))], axis=1)
+    assert not SigMerger.check(
+        [Signal(np.empty((2, 1))), Signal(np.empty((2, 2)))], axis=0)
+
+    # shape mismatch
+    assert not SigMerger.check(
+        [Signal(np.empty((2,))), Signal(np.empty((2, 2)))])
+
+    # mixed dtype
+    assert not SigMerger.check(
+        [Signal(np.empty(2, dtype=int)), Signal(np.empty(2, dtype=float))])
+
+    s1 = Signal(np.empty(5))
+    s2 = Signal(np.empty(5))
+
+    # mixed signal and view
+    assert not SigMerger.check([s1, s1[:3]])
+
+    # mixed bases
+    assert not SigMerger.check([s1[:2], s2[2:]])
+
+    # compatible views
+    assert SigMerger.check([s1[:2], s1[2:]])
+
+
+def test_sigmerger_check_signals():
+    # 0-d signals
+    SigMerger.check_signals([Signal(0), Signal(0)])
+    with pytest.raises(ValueError):
+        SigMerger.check_signals([Signal(0), Signal(1)])
+
+    # compatible along first axis
+    SigMerger.check_signals(
+        [Signal(np.empty((1, 2))), Signal(np.empty((2, 2)))])
+
+    # compatible along second axis
+    SigMerger.check_signals(
+        [Signal(np.empty((2, 1))), Signal(np.empty((2, 2)))], axis=1)
+    with pytest.raises(ValueError):
+        SigMerger.check_signals(
+            [Signal(np.empty((2, 1))), Signal(np.empty((2, 2)))], axis=0)
+
+    # shape mismatch
+    with pytest.raises(ValueError):
+        SigMerger.check_signals(
+            [Signal(np.empty((2,))), Signal(np.empty((2, 2)))])
+
+    # mixed dtype
+    with pytest.raises(ValueError):
+        SigMerger.check_signals(
+            [Signal(np.empty(2, dtype=int)), Signal(np.empty(2, dtype=float))])
+
+    # compatible views
+    s = Signal(np.empty(5))
+    with pytest.raises(ValueError):
+        SigMerger.check_signals([s[:2], s[2:]])
+
+
+def test_sigmerger_check_views():
+    s1 = Signal(np.empty((5, 5)))
+    s2 = Signal(np.empty((5, 5)))
+
+    # compatible along first axis
+    SigMerger.check_views([s1[:1], s1[1:]])
+
+    # compatible along second axis
+    SigMerger.check_views([s1[0:1, :1], s1[0:1, 1:]], axis=1)
```

Comment:
Nitpick: why write `0:1` instead of `:1`. If there's no reason, then I'll make a fixup commit to address it.